Perform a serialization roundtrip on eager apply_async.

charettes opened the pull request:
Fixes #4008 

/cc @AlexHill

AlexHill commented on t/unit/app/test_builtins.py:
```diff
@@ -104,6 +107,11 @@ def test_apply_async_eager(self):
+        self.task.apply = Mock(name='apply')
+        with pytest.raises(EncodeError):
+            self.task.apply_async((1, 2, 3, 4, {1}))
+
```
Is there somewhere better this can go than in the groups test case?

charettes replied:
Probably, I wasn't sure where exactly it should live. Any suggestions?

thedrow replied:
The test case should live under test_tasks.py in my opinion.

AlexHill commented:
Thanks @charettes! @thedrow what's your position on this change?

charettes commented:
> I think we'll need to look at Signature and all of its subclasses as well

Should I move the code the to `Signature.apply_async` instead?

AlexHill commented:
> Should I move the code the to Signature.apply_async instead?

I think having that code where it is now will cover most cases, but I'm not 100% sure, sorry. We'll need to write tests that check that this works consistently with direct invocation (i.e. `my_task.delay(*args)`), signatures (`my_task.s(*args).delay()`), but also chords, chains, groups, etc. I'll look deeper into this when I have more time.

georgepsarakis commented:
@AlexHill I think the change applies for `Signature` as well, since a call to `Signature.apply_async` essentially calls the task's `apply_async`.

@charettes are you interested in continuing this work? I think this is an important change, so it could be included in the next release. Please let me know how we can help.

charettes commented:
@georgepsarakis I rebased and addressed the remaining comment by moving the test to `test_tasks.py`. Let me know if you need anything from me to get this merged.

georgepsarakis commented:
@AlexHill do you have any further feedback on this? Otherwise it should be ready for merge I think.

curtisgr commented:
Hello, the inclusion of this line: https://github.com/celery/celery/pull/4456/files#diff-a4622e17daa3bfaf3dc55dbd6b35ced7R518 changes the behavior such that we now have to specify a broker when running our django test suite, whereas before we did not need to do so.

Is this a desired side effect of this change? Are we abusing celery in some way by not wanting to specify or use a broker during test runs?

alexkiro commented:
> Hello, the inclusion of this line: https://github.com/celery/celery/pull/4456/files#diff-a4622e17daa3bfaf3dc55dbd6b35ced7R518 changes the behavior such that we now have to specify a broker when running our django test suite, whereas before we did not need to do so.

FWIW, I also ran into a situation where this behavior was breaking my tests as well. As I was using various mocked objects that were being passed around through tasks in my tests, and while they were serialized I was losing references to them in the test case. 

To work around this, I disabled this new behavior by registering a `noop` serializer:

```
from kombu import serialization
serialization.register("noop", lambda x: x, lambda x: x, "x-python/noop")
```

And then configuring celery to use that. E.g. for django:

```
from unittest.mock import patch
patch("django.conf.settings.CELERY_TASK_SERIALIZER", "noop").start()
```
